src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala

```diff
@@ -135,34 +139,53 @@ private[impl] class InstanceTrackerActor(
         val op = InstanceOpProcessor.Operation(deadline, sender(), instanceId, instanceUpdateOp)
         updaterRef.forward(InstanceUpdateActor.ProcessInstanceOp(op))
 
-      case msg @ InstanceTrackerActor.StateChanged(ack) =>
-        val maybeChange: Option[InstanceChange] = ack.effect match {
-          case InstanceUpdateEffect.Update(instance, oldInstance, events) =>
-            updateApp(instance.runSpecId, instance.instanceId, newInstance = Some(instance))
-            Some(InstanceUpdated(instance, lastState = oldInstance.map(_.state), events))
+      case InstanceTrackerActor.StateChanged(ack) =>
+        import context.dispatcher
 
-          case InstanceUpdateEffect.Expunge(instance, events) =>
+        val maybeChange: Future[Option[InstanceChange]] = ack.effect match {
+          case effect @ InstanceUpdateEffect.Update(instance, _, _) =>
+            repository.store(instance) // first store to the repository
+              .map(_ => Some(effect))
+              .recoverWith(tryToRecoverRepositoryFailure(effect))
+              .map(trackerEffect => // update task tracked based on the repository update result
+                trackerEffect.flatMap(ef => {
+                  updateApp(ef.instance.runSpecId, ef.instance.instanceId, newInstance = Some(ef.instance))
+                  Some(InstanceUpdated(ef.instance, lastState = ef.oldState.map(_.state), ef.events))
+                }))
+
+          case effect @ InstanceUpdateEffect.Expunge(instance, _) =>
             logger.debug(s"Received expunge for ${instance.instanceId}")
-            updateApp(instance.runSpecId, instance.instanceId, newInstance = None)
-            Some(InstanceDeleted(instance, lastState = None, events))
+            repository.delete(instance.instanceId) // first store to the repository
+              .map(_ => Some(effect))
+              .recoverWith(tryToRecoverRepositoryFailure(effect))
+              .map(trackerEffect => // update task tracked based on the repository update result
+                trackerEffect.flatMap(ef => {
+                  updateApp(ef.instance.runSpecId, ef.instance.instanceId, newInstance = None)
+                  Some(InstanceDeleted(ef.instance, lastState = None, ef.events))
+                }))
 
           case InstanceUpdateEffect.Noop(_) | InstanceUpdateEffect.Failure(_) =>
-            None
+            Future.successful(None)
         }
 
         val originalSender = sender()
 
         import context.dispatcher
-        maybeChange.map { change =>
-          updateStepProcessor.process(change).recover {
-            case NonFatal(cause) =>
-              // since we currently only use ContinueOnErrorSteps, we can simply ignore failures here
-              logger.warn("updateStepProcessor.process failed", cause)
-              Done
-          }
-        }.getOrElse(Future.successful(Done)).foreach { _ =>
-          ack.sendAck()
-          originalSender ! (())
+        maybeChange.flatMap { changeOpt =>
```

Comment:
again, I'd prefer to see await/async here - it makes it explicit that the code beneath now happens in a different context.